File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    if number <= 0:
        raise ValueError("Input must be a positive number")

    factors = [x for x in range(1, int(number)) if number % x == 0]
    aliquot_sum = sum(factors)

    if aliquot_sum == number:
        return "perfect"
    elif aliquot_sum < number:
        return "deficient"
    else:
        return "abundant"

File: scala/perfect-numbers/src/main/scala/PerfectNumbers.scala
sealed trait Classification

object NumberType {
  case object Perfect extends Classification
  case object Abundant extends Classification
  case object Deficient extends Classification
}

object PerfectNumbers {
  def classify(n: Int): Either[String, Classification] = {
    if (n <= 0) {
      return Left("Classification is only possible for natural numbers.")
    }

    val factors = Range.inclusive(1, n / 2).filter { n % _ == 0 }

    val classification = factors.sum match {
      case x if x == n => NumberType.Perfect
      case x if x < n => NumberType.Deficient
      case x if x > n => NumberType.Abundant
    }

    Right(classification)
  }
}